### PR TITLE
Add edge-offline commands

### PR DESCRIPTION
**Pull request by anusapan**
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** unit **and** integration tests passed locally? i.e. `pytest -s <project root>/azext_iot/tests`
- [ ] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.

### CODE REVIEW

File: azext_iot/_help.py [hunk 1]
```diff
@@ -518,6 +518,57 @@
     short-summary: Upload a local file as a device to a pre-configured blob storage container.
 """
 
+helps['iot device get-parent'] = """
+    type: command
+    short-summary: Get the parent device of the specified device.
+    examples:
+    - name: Get the parent device of the specified device .
```
Comment (YingXue) [inline azext_iot/_help.py]: unnecesary white space before period.

File: azext_iot/_help.py [hunk 2]
```diff
@@ -518,6 +518,57 @@
...
+      text: >
+        az iot device get-parent -d [Non_Edge_Device_Id] -n [IoTHub_Name]
+"""
+
+helps['iot device children'] = """
+    type: group
+    short-summary: Manage Edge offline.
+"""
+
+helps['iot device children add'] = """
```
Comment (YingXue) [inline azext_iot/_help.py]: I am not sure if the spec got updated or not, I remember we need to implement children SET command as well? I could be wrong, can you double check with Venkat?

File: azext_iot/_help.py [hunk 3]
```diff
@@ -518,6 +518,57 @@
...
+    short-summary: Print comma-separated list of assigned child devices.
+    examples:
+    - name: Show all assigned non-edge devices as comma-separated list.
+      text: >
+        az iot device children list -d [Edge_Device_Id] -n [IoTHub_Name]
+"""
+
+helps['iot device children remove'] = """
+    type: command
+    short-summary: Remove comma-separated list of devices as children of specified device.
```
Comment (YingXue) [inline azext_iot/_help.py]: Maybe rephrase this short summary because users don't need to specify a list if they do --remove-all
Can be "Remove non edge devices as children from specified edge device" instead.

File: azext_iot/_params.py [hunk 4]
```diff
@@ -206,6 +206,22 @@ def load_arguments(self, _):
         context.argument('content_type', options_list=['--content-type', '--ct'],
                          help='MIME Type of file.')
 
+    with self.argument_context('iot device get-parent') as context:
+        context.argument('device_id', help='This is an non-Edge device.')
```
Comment (YingXue) [inline azext_iot/_params.py]: Rephrase it to 'Id of non-edge device'. I also notice sometimes we use Edge and sometimes edge. Can you find out what we currently use in existing commands and be consistent here?

File: azext_iot/_params.py [hunk 5]
```diff
@@ -206,6 +206,22 @@ def load_arguments(self, _):
...
+
+    with self.argument_context('iot device children add') as context:
+        context.argument('device_id', help='This is an Edge device.')
+
+    with self.argument_context('iot device children add') as context:
+        context.argument('child_list', options_list=['--child-list'],
+                         help='Child device list includes only non-Edge devices.')
+        context.argument('force', options_list=['--force', '-f'], help='Overwrites the non-Edge device\'s parent set.')
```
Comment (YingXue) [inline azext_iot/_params.py]: parent set?

File: azext_iot/operations/hub.py [hunk 6]
```diff
@@ -181,6 +181,114 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
         raise CLIError(unpack_msrest_error(e))
 
 
+def iot_device_get_parent(cmd, device_id, hub_name=None, resource_group_name=None, login=None):
+    target = get_iot_hub_connection_string(cmd, hub_name, resource_group_name, login=login)
+    child_device = _iot_device_show(target, device_id)
+    if child_device['capabilities']['iotEdge']:
+        raise CLIError('The device should be non-Edge device')
+    if 'deviceScope' not in child_device:
+        raise CLIError('The device doesn\'t have any parent set')
```
Comment (YingXue) [inline azext_iot/operations/hub.py]: Wording here is quite confusing. 
Maybe change it to 'Device <id> doesn't support parent device functionality'

File: azext_iot/operations/hub.py [hunk 7]
```diff
@@ -181,6 +181,114 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
...
+        nonedge_device = _iot_device_show(target, non_edge_device_id.strip())
+        _validate_nonedge_device(nonedge_device)
+        if 'deviceScope' not in nonedge_device or nonedge_device['deviceScope'] == '':
+            devices.append(nonedge_device)
+            continue
+        if nonedge_device['deviceScope'] != edge_device['deviceScope']:
+            if force:
+                devices.append(nonedge_device)
+                continue
+            raise CLIError('The entered device "{}" already has a parent set and use \'--force\''
```
Comment (YingXue) [inline azext_iot/operations/hub.py]: Device "{}" already has a parent device, please use \'--force\'

File: azext_iot/operations/hub.py [hunk 8]
```diff
@@ -181,6 +181,114 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
...
+    elif child_list:
+        edge_device = _iot_device_show(target, device_id)
+        _validate_edge_device(edge_device)
+        for non_edge_device_id in child_list.split(','):
+            nonedge_device = _iot_device_show(target, non_edge_device_id.strip())
+            _validate_nonedge_device(nonedge_device)
+            if 'deviceScope' not in nonedge_device or nonedge_device['deviceScope'] == '':
+                raise CLIError('The entered device "{}" isn\'t a child device.'.format(non_edge_device_id.strip()))
+            if nonedge_device['deviceScope'] != edge_device['deviceScope']:
+                raise CLIError('The entered child device "{}" isn\'t assigned a child of edge device "{}"'
```
Comment (YingXue) [inline azext_iot/operations/hub.py]: The entered child device "{}" isn\'t assigned as a child of edge device

File: azext_iot/_help.py (hunk 1 above)
Comment (anusapan) [reply azext_iot/_help.py]: Fixed

File: azext_iot/_help.py (hunk 3 above)
Comment (anusapan) [reply azext_iot/_help.py]: Fixed

File: azext_iot/_params.py (hunk 4 above)
Comment (anusapan) [reply azext_iot/_params.py]: Fixed

File: azext_iot/_params.py (hunk 5 above)
Comment (anusapan) [reply azext_iot/_params.py]: Fixed

File: azext_iot/operations/hub.py (hunk 7 above)
Comment (anusapan) [reply azext_iot/operations/hub.py]: Fixed

File: azext_iot/operations/hub.py (hunk 8 above)
Comment (anusapan) [reply azext_iot/operations/hub.py]: Fixed

File: azext_iot/tests/test_iot_ext_int.py [hunk 9]
```diff
@@ -1399,3 +1399,116 @@ def test_storage(self):
                      self.check('failureReason', None),
                      self.check('type', 'export'),
                      self.exists('jobId')])
+
```
Comment (YingXue) [inline azext_iot/tests/test_iot_ext_int.py]: do we not have unit test for these commands?

File: azext_iot/_params.py [hunk 10]
```diff
@@ -206,6 +206,22 @@ def load_arguments(self, _):
...
+    with self.argument_context('iot device children add') as context:
+        context.argument('device_id', help='Id of edge device.')
+
+    with self.argument_context('iot device children add') as context:
+        context.argument('child_list', options_list=['--child-list'],
+                         help='Child device list includes only non-edge devices.')
+        context.argument('force', options_list=['--force', '-f'], help='Overwrites the non-edge device\'s parent device.')
+
+    with self.argument_context('iot device children remove') as context:
+        context.argument('child_list', options_list=['--child-list'],
```
Comment (digimaun) [inline azext_iot/_params.py]: The options list for parameters should include short aliases (like --cl or similar) unless there is a reason the user should spell the whole thing out.

File: azext_iot/tests/test_iot_ext_int.py (hunk 9 above)
Comment (digimaun) [reply azext_iot/tests/test_iot_ext_int.py]: +1, unit tests please.

File: azext_iot/_params.py (hunk 10 above)
Comment (anusapan) [reply azext_iot/_params.py]: done

File: azext_iot/tests/test_iot_ext_int.py (hunk 9 above)
Comment (anusapan) [reply azext_iot/tests/test_iot_ext_int.py]: added

File: azext_iot/_help.py (hunk 2 above)
Comment (anusapan) [reply azext_iot/_help.py]: Not required, as remove-all and add will be "set" functionality

File: azext_iot/operations/hub.py [hunk 11]
```diff
@@ -181,6 +181,112 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
         raise CLIError(unpack_msrest_error(e))
 
 
+def iot_device_get_parent(cmd, device_id, hub_name=None, resource_group_name=None, login=None):
+    target = get_iot_hub_connection_string(cmd, hub_name, resource_group_name, login=login)
+    child_device = _iot_device_show(target, device_id)
+    if child_device['capabilities']['iotEdge']:
```
Comment (digimaun) [inline azext_iot/operations/hub.py]: Why not use one of the new validate helpers below?

File: azext_iot/commands.py [hunk 12]
```diff
@@ -87,6 +87,12 @@ def load_command_table(self, _):
         cmd_group.command('send-d2c-message', 'iot_device_send_message')
         cmd_group.command('simulate', 'iot_simulate_device')
         cmd_group.command('upload-file', 'iot_device_upload_file')
+        cmd_group.command('get-parent', 'iot_device_get_parent')
```
Comment (digimaun) [inline azext_iot/commands.py]: This is not an implementation comment. This is a requirements/design comment.

Right now I am not so sure if it makes sense for these commands to be under the `iot device` sub group when we have `iot hub device-*` and `iot edge` command groups which respectively map to device manipulation (like device identity, device twin) and edge deployment management.

These new commands are aligned to IoT Edge and as it is today `iot device` command group are operations 'as device' i.e. messaging, simulation, file upload.

See issue [#46](https://github.com/Azure/azure-iot-cli-extension/issues/46#issuecomment-425228047) for some additional details.

The core point here is the command structure needs to be organized and logically aligned. If today's structure is not intuitive then it should be improved/reformulated (more on that in the linked issue).

Comment (jongio) [reply azext_iot/commands.py]: @iluican and @wickste for comments.

File: azext_iot/operations/hub.py (hunk 11 above)
Comment (anusapan) [reply azext_iot/operations/hub.py]: Fixed

File: azext_iot/operations/hub.py [hunk 13]
```diff
@@ -181,6 +202,127 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
...
+        edge_device = _iot_device_show(target, device_id)
+        _validate_edge_device(edge_device)
+        for non_edge_device_id in child_list.split(','):
+            nonedge_device = _iot_device_show(target, non_edge_device_id.strip())
+            _validate_nonedge_device(nonedge_device)
+            _validate_child_device(nonedge_device)
+            if nonedge_device['deviceScope'] != edge_device['deviceScope']:
+                raise CLIError('The entered child device "{}" isn\'t assigned as a child of edge device "{}"'
+                               .format(non_edge_device_id.strip(), device_id))
+            if nonedge_device['deviceScope'] == edge_device['deviceScope']:
```
Comment (digimaun) [inline azext_iot/operations/hub.py]: Do you need this `if`, seems like this check can be an `else`....then remove the `continue` couple lines down since there is no additional work being done.

File: azext_iot/operations/hub.py [hunk 14]
```diff
@@ -181,6 +202,127 @@ def iot_device_delete(cmd, device_id, hub_name=None, resource_group_name=None, l
...
+            service_sdk.create_or_update_device(nonedge_device['deviceId'], nonedge_device, custom_headers=headers)
+        else:
+            raise LookupError("device etag not found.")
+    except errors.CloudError as e:
+        raise CLIError(unpack_msrest_error(e))
+
+
+def _validate_edge_device(device):
+    if not device['capabilities']['iotEdge']:
+        raise CLIError('The device should be edge device.')
```
Comment (digimaun) [inline azext_iot/operations/hub.py]: It certainly helps user experience when you add additional detail to the error, in this instance, addition text to call out the device Id failing this check (like you have in _validate_nonedge_device).

Comment (anusapan) [reply azext_iot/operations/hub.py]: Fixed

File: azext_iot/operations/hub.py (hunk 13 above)
Comment (anusapan) [reply azext_iot/operations/hub.py]: Fixed
